### ISSUE TITLE
Add tests and automated self-healing for corrupted cache states (binaries, annotations, metadata, and legacy cache)

### ISSUE THREAD

**Issue by vugi99** [vugi99/nanos-lint#6]
### Summary

Audit and improve error handling, meaningful diagnostics, and automated self-healing mechanisms across `nanos-lint` when encountering corrupted, truncated, or older/legacy cache states:
- Corrupted or 0-byte `annotations.lua` in user cache.
- Corrupted or truncated `lua-language-server` binary (especially when operating offline).
- Malformed or invalid `metadata.json` files for both LuaLS and annotations.
- Corrupted or incomplete older/legacy cache states (`nanos-lint <= 2.2.1`).

Expand the test suite with dedicated unit and integration tests simulating these failure states to guarantee clean recovery or actionable diagnostics rather than cryptic crashes or false-positive lint errors.

---

### Current Test Suite & Codebase Audit

An audit of the existing codebase (`src/luals.ts`, `src/annotations.ts`) and test suites (`tests/unit/luals-cache.test.ts`, `tests/unit/annotations.test.ts`, `tests/unit/regressions.test.ts`) reveals the following current behaviors:

#### 1. LuaLS Binary Validation
- **What is implemented**: `isBinaryValid(binaryPath)` verifies that the binary exists, has a size $\ge 100,000$ bytes, and executes `--version` returning SemVer output within 5 seconds.
- **Current Tests**:
  - `tests/unit/regressions.test.ts`: Verifies `isBinaryValid` rejects fake truncated binaries, and that `downloadAndExtractLuaLS` repairs target directories when online.
  - `tests/unit/luals-cache.test.ts`: Verifies `.complete` marker mismatches trigger invalidation.
- **Identified Failure Mode**:
  - If a cached LuaLS binary is corrupted or truncated and the user is **offline** or has network timeouts, `resolveLuaLSBinary` purges the corrupted directory and immediately tries to fetch over the network, failing with a raw network error (`TypeError: fetch failed`). The user never learns that their cached binary was corrupted or how to fix it.

#### 2. Annotations Validation (`annotations.lua`) — *Critical Blind Spot*
- **What is implemented**:
  - `fetchRawAnnotationsContent` validates size upon download (`text.length >= 1000`).
  - However, `resolveAnnotations` **never validates the size or content of an already cached `annotations.lua` file**!
  - Lines 325 and 344 in `src/annotations.ts` only check:
    ```typescript
    if (metadata && metadata.lastChecked === dateStr && fs.existsSync(cachedAnnotationsFile)) {
      return cachedAnnotationsFile;
    }
    ```
- **Identified Failure Mode**:
  - If `annotations.lua` is truncated to 0 bytes (e.g. power loss, disk full, interrupted process, or manual edit), `resolveAnnotations` considers it valid because the file exists.
  - LuaLS is then invoked with an empty definitions file, causing every single nanos world API call (`Character`, `Vehicle`, `Player`, `Events`, etc.) to fail resolution and flood the user with hundreds of false-positive diagnostics.
  - The cache will not auto-heal until the next calendar day (or until the user manually runs `nanos-lint clean-cache`).
  - When offline with a truncated `annotations.lua`, `resolveAnnotations` also returns the empty file without validation.

#### 3. Metadata Files (`metadata.json`)
- **What is implemented**:
  - `readLuaLSMetadata` and `readAnnotationsMetadata` return `null` when JSON parsing fails or required fields are missing.
  - Both have basic unit tests asserting `null` return on `"invalid-json"`.
- **Identified Failure Mode**:
  - When `metadata.json` contains malformed JSON, neither module automatically repairs or resets the invalid file on disk until a subsequent network download succeeds. If offline, the corrupted JSON file persists indefinitely.

#### 4. Legacy Cache Migration (`nanos-lint <= 2.2.1`)
- **What is implemented**:
  - `findExistingLuaLSDir` and `resolveLuaLSBinary` probe `getLegacyCacheDir(version)`. If `isBinaryValid` passes, they copy files from legacy to new cache paths.
- **Current Tests / Gaps**:
  - There are currently no tests asserting that an incomplete or corrupted legacy directory is safely ignored without attempting migration or crashing on partial copy.

---

#### 1. Introduce `isAnnotationsValid(filePath)`
Add a validator function in `src/annotations.ts`:
- Check file exists, is a regular file, and `stat.size >= 1000` bytes.
- Check initial content contains expected nanos world header/signatures (e.g., `--` or `nanos world`).
- If invalid or 0 bytes, `resolveAnnotations` must:
  1. Log a warning: `[annotations] Cached annotations.lua is corrupted or empty. Purging and refreshing...`.
  2. Delete the corrupted cached file.
  3. Attempt to re-download.
  4. If offline, fall back to bundled `annotations.lua` from package root instead of using a broken file.

#### 2. Actionable Offline Diagnostics for Corrupted LuaLS Binaries
In `resolveLuaLSBinary`:
- If an existing cached binary fails `isBinaryValid` and network re-download fails:
  - Throw a clear, human-actionable error:
    `"Cached LuaLS binary at '<path>' is corrupted (failed execution/size check) and cannot be re-downloaded while offline. Please connect to the internet to repair or run 'nanos-lint clean-cache'."`

#### 3. Self-Healing Corrupted Metadata
- When `readLuaLSMetadata` or `readAnnotationsMetadata` catches a JSON parse error, automatically remove or overwrite the corrupted `metadata.json` with a clean state rather than leaving corrupted JSON on the filesystem.

---

### Planned Test Suite Additions

Implement dedicated test scenarios in `tests/unit/`:

1. **`tests/unit/annotations-corruption.test.ts` (or additions to `annotations.test.ts`)**:
   - **0-byte cached `annotations.lua`**: Verify `resolveAnnotations` detects the 0-byte file, discards it, and recovers by re-fetching or falling back to bundled annotations.
   - **Corrupted cached `annotations.lua` while offline**: Verify that an invalid cache does not get passed to the linter; instead, it safely falls back to bundled annotations.
   - **Malformed `metadata.json`**: Verify that corrupt metadata is cleaned up and self-heals upon next run.

2. **`tests/unit/luals-corruption.test.ts` (or additions to `luals-cache.test.ts`)**:
   - **Corrupted binary + offline mode**: Verify that attempting to resolve a corrupted binary with network failure throws a descriptive corruption error pointing to the cache directory and remediation steps.
   - **Legacy cache with truncated binary**: Verify that `findExistingLuaLSDir` ignores a corrupted legacy cache directory and does not copy corrupted files into the primary cache.
   - **Malformed LuaLS `metadata.json` while offline**: Verify that fallback to existing valid version directories still succeeds even if `metadata.json` contains corrupted JSON.

---

### Action Items

- [ ] Implement `isAnnotationsValid(filePath: string): boolean` in `src/annotations.ts`.
- [ ] Add auto-healing fallback to bundled annotations when cached annotations are truncated/corrupt.
- [ ] Enhance error messaging in `resolveLuaLSBinary` when offline repair fails.
- [ ] Write unit tests for 0-byte annotations recovery, offline binary corruption errors, and legacy cache rejection.
- [ ] Verify all quality gates pass (`npm run lint`, `npm run typecheck`, `npm run build`, `npm run test:coverage`).
